ipc_service: icmsg: Improve packets reception

KAGA164 opened the pull request:
This improves packets reception in the icmsg backend. The remote can receive the same amount of data that were send by host. It removes local data coping into the internal buffer and fix and issue where data from host larger than remote buffer was silently dropped. Additionally this PR contains documentation clarification for the spsc_pbuf_claim function.

Commit:
ipc: icmsg: Improve packets reception

This improves packet reception and fix an issue
where packets bigger than internal Rx buffer
were silently dropped. In current solution
local data coping for reception is not needed.
User gets direct pointer to shared memory which
allow to efficient receive as much data as was sent.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

Commit:
os: spsc_buf: Clarify claim function doc

This clarifies the spsc_buf_claim function documentation.
It should be easier now to understand what this function
does and what are returned memory properties.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

emob-nordic commented on subsys/ipc/ipc_service/lib/icmsg.c:
```diff
@@ -159,31 +162,47 @@ static void submit_work_if_buffer_free_and_data_available(
-				 RX_BUF_SIZE);
 
-	__ASSERT_NO_MSG(len <= RX_BUF_SIZE);
+	uint16_t len = spsc_pbuf_claim(dev_data->rx_ib, &rx_buffer);
```
When i first saw `spsc_pbuf_claim` i thaught this cannot be done, as the user could call https://github.com/zephyrproject-rtos/zephyr/blob/df0e03c5512c03f4e3e641cc37641abb1f76082b/include/zephyr/ipc/ipc_service_backend.h#L199 where surely `spsc_pbuf_claim` is called, well - its not. This surprises me, i will review this later just tossing this out out as it seems odd we have different mechanism for claiming rx buffer when no copy rx is used.

hubertmis commented on subsys/ipc/ipc_service/lib/icmsg.c:
```diff
@@ -159,31 +162,47 @@ static void submit_work_if_buffer_free_and_data_available(
+			/* Release Rx buffer here only in case when user did not request
+			 * to hold it.
+			 */
+			if (is_rx_buffer_free(dev_data)) {
```
```suggestion
			if (!is_rx_buffer_held(dev_data)) {
```
would be easier to read

KAGA164 replied:
Is it make sense to you to create the new static function for opposite condition checking?

hubertmis replied:
Yes, it improves code readability. You are interested here if the callback is requested to hold the buffer. Not if it is "free" whatever it means in this context. This change would clearly show the programmer's intention. While the assembly should be exactly the same after optimization.

KAGA164 replied:
Ok, done

hubertmis reviewed (dismissed):
Looks good

KAGA164 commented on subsys/ipc/ipc_service/lib/icmsg.c:
```diff
@@ -469,12 +487,21 @@ int icmsg_release_rx_buffer(const struct icmsg_config_t *conf,
+	dev_data->rx_buffer = NULL;
+	dev_data->rx_len = 0;
+
+	atomic_set(&dev_data->rx_buffer_state, RX_BUFFER_STATE_RELEASED);
```
The atomic_x operations can be implemented differently for various architectures. Mostly it is generic Zephyr implementation where spinlock is used so basically it is a irq lock/unlcok under. But the issue might be there because for ARM V7/V8 M there is synchronization instruction to ensure that lock or unlock is effective immediately, for RISCV it looks like we don't use barrier so there might be not guaranty that irq_lock/unlock will be executed in requested order. Anyway I am not an expert in our VPRs architecture. I think this possible issue should be solved under atomic implementation.

hubertmis replied:
Fully agree it should be solved in atomics API and their implementations. Not in this PR

Commit:
ipc: icmsg: Align naming for buffers

This remanes send_buffer to tx_buffer to be
consistent with Rx buffer naming and with
mbox naming.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

KAGA164 commented:
Just rebase plus small naming change send_buffer->tx_buffer

emob-nordic commented:
Just come comments to consider.

carlocaione commented:
@emob-nordic can you take ownership of `subsys/ipc/ipc_service/lib/icmsg.c` as well in the `CODEOWNERS` file?

emob-nordic commented on subsys/ipc/ipc_service/lib/icmsg.c:
```diff
@@ -70,54 +73,72 @@ static bool is_endpoint_ready(struct icmsg_data_t *dev_data)
+
+#if CONFIG_IPC_SERVICE_ICMSG_NOCOPY_RX
+	dev_data->rx_buffer = NULL;
+	dev_data->rx_len = 0;
+#endif
```
We should set the state here, shouldn't we?

``` suggestion
	dev_data->rx_len = 0;
	atomic_set(&dev_data->rx_buffer_state, RX_BUFFER_STATE_RELEASED);
#endif
}
```

KAGA164 replied:
Discussed offline. It cannot be wrapped in clear way in static function

emob-nordic commented on subsys/ipc/ipc_service/lib/icmsg.c:
```diff
@@ -70,54 +73,72 @@ static bool is_endpoint_ready(struct icmsg_data_t *dev_data)
 #endif
 }
 
+static bool is_rx_buffer_held(struct icmsg_data_t *dev_data)
```
Grammatically correct this should be hold, not held.

``` suggestion
static bool is_rx_buffer_hold(struct icmsg_data_t *dev_data)
```

KAGA164 replied:
The same here

eobalski reviewed (approved):
Concerns discussed offline. lgtm
